whois-api/src/main/java/net/ripe/db/whois/api/mail/dequeue/BouncedMessageParser.java

```diff
@@ -0,0 +1,132 @@
+package net.ripe.db.whois.api.mail.dequeue;
+
+import jakarta.mail.Address;
+import jakarta.mail.MessagingException;
+import jakarta.mail.Part;
+import jakarta.mail.internet.AddressException;
+import jakarta.mail.internet.ContentType;
+import jakarta.mail.internet.InternetAddress;
+import jakarta.mail.internet.InternetHeaders;
+import jakarta.mail.internet.MimeMessage;
+import net.ripe.db.whois.api.mail.MessageInfo;
+import org.eclipse.angus.mail.dsn.DeliveryStatus;
+import org.eclipse.angus.mail.dsn.MultipartReport;
+import org.eclipse.angus.mail.dsn.Report;
+import org.elasticsearch.common.Strings;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+@Component
+public class BouncedMessageParser {
+
+    private static final ContentType MULTIPART_REPORT = contentType("multipart/report");
+
+    private final boolean enabled;
+
+    @Autowired
+    public BouncedMessageParser(@Value("${mail.smtp.from:}") final String smtpFrom) {
+        this.enabled = !Strings.isNullOrEmpty(smtpFrom);
+    }
+
+    @Nullable
+    public MessageInfo parse(final MimeMessage message) throws MessagingException, IOException {
+        if (enabled && isMultipartReport(message)) {
+            final MultipartReport multipartReport = multipartReport(message.getContent());
+            if (isReportDeliveryStatus(multipartReport)) {
+                final DeliveryStatus deliveryStatus = deliveryStatus(message);
+                if (isFailed(deliveryStatus)) {
+                    final MimeMessage returnedMessage = multipartReport.getReturnedMessage();
+                    final String messageId = getMessageId(returnedMessage.getMessageID());
+                    // TODO: double check we have the right recipient (This is the TO: header)
```

Comment:
Remove unnecessary TODO, we always only have one recipient, take the first one. The validation of the message id + email address happens later.